### PR TITLE
add READ message handling

### PR DESCRIPTION
**Pull request by lumisphere902**
- Add read message handling
- Sends all other messages to connections which are "reading" from the source (currently only the teacher)
- Updates will be handled and sent to the recipient separately

### CODE REVIEW

File: db/collab.go
Comment (krashanoff) [inline db/collab.go:103]: This is more semantically clear:

```suggestion
func (s *Session) BroadcastTo(msg Message, uids ...string) (lastErr error) {
	s.Lock.Lock()
	defer s.Lock.Unlock()
	for _, uid := range uids {
		if err := websocket.JSON.Send(s.Conns[uid], msg); err != nil {
			lastErr = err
		}
	}
	return
}
```

File: db/collab.go
Comment (krashanoff) [inline db/collab.go:27]: What do you think about extending the websocket `Conn` type with our own fields to simplify the code a bit? It would allow set broadcast functions to be localized to per-conection bases. Here's a suggested starter for the signature:

```
type Connection struct {
    *websocket.Conn
    UID string
    Subscriptions []string
}

type Session struct {
	// Map UIDs to their websocket.Conn
	Conns   map[string]Connection
```

I'm not too particular either way, just wanted to field some thoughts.

File: db/collab.go
Comment (krashanoff) [inline db/collab.go:227]: The rightward pressure here is too great. I think this packet management should be abstracted to a member function of our `Session` type. Alternatively, the introduction of our extension of the `Conn` type could solve this issue, too.

File: db/collab.go
Comment (krashanoff) [inline db/collab.go:198]: You can just reinitialize and declare `err` here. These vars are scope-local.

```suggestion
				if err := session.BroadcastError(uid, err.Error()); err2 != nil {
```

**Review by krashanoff — CHANGES_REQUESTED** [review]
Hey Timothy, apologies on getting this review out late. I have left some feedback on your changes. The biggest concern at present is the organization of the code. There are opportunities to make it more idiomatic and save some headaches for future maintainers.

File: db/collab.go
Comment (lumisphere902) [reply db/collab.go:227]: Can you elaborate on what exactly should be abstracted here? Sorry, not entirely sure what you're referring to

File: db/collab.go
Comment (krashanoff) [inline db/collab.go:94]: While wrapping up documentation, two short doc comments on `BroadcastTo` and `BroadcastToSet` will help.

**Review by krashanoff — CHANGES_REQUESTED** [review]
Hey Timothy, I'm liking the changes you've made for code clarity. The `StringSet` type probably doesn't need to be exported, but I think it's solid nonetheless.

Almost there, but there are a few minor adjustments that should be made before it's cleared. In the meantime, I'll see if I can do some testing to confirm functionality before merging. This can be put into a separate PR, but here's a great article on [WebSocket tests in Go](https://quii.gitbook.io/learn-go-with-tests/build-an-application/websockets) for automating the process.

File: db/collab.go
Comment (lumisphere902) [reply db/collab.go:127]: Sorry, what exactly do you mean by this?

Comment (krashanoff) [reply db/collab.go:127]: I meant it in the parallel sense. Since this function mutates the actual structure, we should lock the data type. Something like:

```go
func (s *Session) RequestAccess(uid string, msg Message) error {
	s.Lock()
	defer s.Unlock()
	if msg.Author == s.Teacher {
		// ...
```

Will do the trick.

**Review by krashanoff — APPROVED** [review]
Looks good!

**Pull request merged**
